middleware/rollup.go

```diff
@@ -0,0 +1,164 @@
+	l := &RollupRouter{
+		logger:  logger,
+		rollups: map[string]*logRollup{},
+		ctx:     ctx,
+		ctxDone: false,
+	}
+	go func() {
+		select {
+		case <-ctx.Done():
+			l.rollupsMu.Lock()
+			l.rollups = map[string]*logRollup{}
+			l.ctxDone = true
+			l.rollupsMu.Unlock()
+		}
+	}()
+	return l
+}
+
+// Process a log message to roll up.
+func (r *RollupRouter) Process(logmsg map[string]interface{}) {
+	if r.ctxDone {
+		return
+	}
+
+	statusCode, ok := logmsg["status-code"].(int)
+	if !ok {
+		return
+	}
+	path, ok := logmsg["path"].(string)
+	if !ok {
+		return
+	}
+	canary, ok := logmsg["canary"].(bool)
+	if !ok {
+		return
+	}
+	r.findOrCreate(statusCode, path, canary).add(logmsg)
+}
+
+func (r *RollupRouter) findOrCreate(statusCode int, path string, canary bool) *logRollup {
+	r.rollupsMu.Lock()
+	defer r.rollupsMu.Unlock()
+	rollupKey := fmt.Sprintf("%d-%s", statusCode, path)
+	if canary {
+		rollupKey += "-canary"
+	}
+	if rollup, ok := r.rollups[rollupKey]; ok {
+		return rollup
+	}
+	rollup := &logRollup{
+		Logger:           r.logger,
+		StatusCode:       statusCode,
+		Path:             path,
+		Canary:           canary,
+		ReportingDelayNs: (time.Second * 20).Nanoseconds(), // todo: make configurable
+	}
+	r.rollups[rollupKey] = rollup
+	go rollup.schedule(r.ctx)
+	return rollup
+}
+
+// logRollup represents a single rollup.
+type logRollup struct {
+	Logger           RollupLogger
+	StatusCode       int
+	Path             string
+	Canary           bool
+	ReportingDelayNs int64
+
+	rollupMsgMu sync.Mutex
+	rollupMsg   map[string]interface{}
+}
+
+func (r *logRollup) report() {
+	r.rollupMsgMu.Lock()
+	defer r.rollupMsgMu.Unlock()
+	if r.rollupMsg != nil {
+		switch logLevelFromStatus(r.StatusCode) {
+		case logger.Error:
+			r.Logger.ErrorD("request-finished", r.rollupMsg)
+		default:
+			r.Logger.InfoD("request-finished", r.rollupMsg)
+		}
+	}
+	r.rollupMsg = nil
+}
+
+func (r *logRollup) schedule(ctx context.Context) {
+	lastReport := time.Now()
+	for {
+		reportingDelay := time.Duration(atomic.LoadInt64(&r.ReportingDelayNs))
+		wakeupTime := lastReport.Add(reportingDelay)
+		now := time.Now()
+		if now.After(wakeupTime) {
+			wakeupTime = now.Add(reportingDelay)
+		}
+		sleepTime := wakeupTime.Sub(now)
+
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(sleepTime):
+			lastReport = time.Now()
+			r.report()
+		}
+	}
+}
+
+func (r *logRollup) add(logmsg map[string]interface{}) {
+	r.rollupMsgMu.Lock()
+	defer r.rollupMsgMu.Unlock()
+
+	if r.rollupMsg == nil {
+		r.rollupMsg = map[string]interface{}{
+			"status-code": r.StatusCode,
+			"path":        r.Path,
+			"count":       int64(0),
+			"canary":      r.Canary,
+			"via":         "kayvee-middleware",
+		}
+	}
+
+	r.rollupMsg["count"] = r.rollupMsg["count"].(int64) + 1
+	// TODO: allow users to perform additional numeric rollups, e.g. if there's a field `response-time`, let them average it
```

Comment:
As for v1, I'd only add response time rollups if app-service (or wherever we plan to test) cares about monitoring that data.

I do think we should add it in the not distant future.

Reply:
Mainly, I suggest that we make it feature complete before doing a rollout to many apps, so we dont have to do that rollout multiple times.

Reply:
added response time rollup 509e68e